README.md
# PESetEnv [![Flattr this][flatter_png]][flatter]

Set environment in [BartPE][].

## Usage

See plugin\z_pesetenv\z_pesetenv.htm for details

## Verify a Release

To verify a release, download the .zip, .sha256, and .asc files for the release 
(replacing PESetEnv-1.4-win32.zip with the release you are verifying):

````
$ wget https://github.com/rasa/PESetEnv/releases/download/v1.4/PESetEnv-1.4-win32.zip{,.sha256,.asc}
````

Next, check that sha256sum reports "OK":
````
$ sha256sum -c PESetEnv-1.4-win32.zip.sha256
PESetEnv-1.4-win32.zip: OK
````

Lastly, check that GPG reports "Good signature":

````
$ gpg --keyserver hkps.pool.sks-keyservers.net --recv-key 0x105a5225b6ab4b22
$ gpg --verify PESetEnv-1.4-win32.zip.asc PESetEnv-1.4-win32.zip
gpg:                using RSA key 0xFF914F74B4BB6EF3
gpg: Good signature from "Ross Smith II <ross@smithii.com>" [ultimate]
...
````

## Contributing

To contribute to this project, please see [CONTRIBUTING.md](CONTRIBUTING.md).

## Bugs

To view existing bugs, or report a new bug, please see [issues](../../issues).

## Changelog

To view the version history for this project, please see [CHANGELOG.md](CHANGELOG.md).

## License

This project is [MIT licensed](LICENSE).

## Contact

This project was created and is maintained by [Ross Smith II][] [![endorse][endorse_png]][endorse]

Feedback, suggestions, and enhancements are welcome.

[Ross Smith II]: mailto:ross@smithii.com "ross@smithii.com"
[flatter]: https://flattr.com/submit/auto?user_id=rasa&url=https%3A%2F%2Fgithub.com%2Frasa%2FPESetEnv
[flatter_png]: http://button.flattr.com/flattr-badge-large.png "Flattr this"
[endorse]: https://coderwall.com/rasa
[endorse_png]: https://api.coderwall.com/rasa/endorsecount.png "endorse"

[BartPE]: http://www.nu2.nu/pebuilder/

CHANGELOG.md
## 1.6 (unreleased)

  * Update Makefile
  
## 1.5 (28-Mar-2015)

  * Fix segfault due to linker issue (#1)

## 1.4 (27-Mar-2015)

  * Executables signed with StartSSL code signing certificate
  * Release zip files are gpg signed and sha256 hashed
  * Standardize build scripts
  * Update documentation

## 1.3 (22-Feb-2015)

  * Updated to Visual Studio 2013
  * Switch to MIT license

## 1.2 (19-Jul-2006)

  * Standardized version information across all programs

## 1.1 (08-Apr-2006)

  * Added .rc and .ico files
  * Improved build scripts
  * Updated documentation

## 1.0 (14-Jul-2005)

  * Initial release


PESetEnv.cpp
// Copyright (c) 2005-2015 Ross Smith II. See Mit LICENSE in /LICENSE

#ifdef _MSC_VER
#pragma warning(disable:4996)
#pragma comment(lib, "advapi32.lib")
#pragma comment(lib, "shell32.lib")
#pragma comment(lib, "user32.lib")
#endif

#ifndef WIN32_LEAN_AND_MEAN
# define WIN32_LEAN_AND_MEAN 1
#endif

#define _WIN32_WINNT 0x501

#include <windows.h>
#include <stdio.h>
#include <stdlib.h>
//#include <atlbase.h> // required for Stackwalker
#include <shellapi.h>
#include <tchar.h>

#include "SystemMessage.h"
#include "debug.h"
#include "UpdatePathVariable.h"
#include "getopt.h"

#include "version.h"

#define APPNAME			L"PESetEnv"
#define APPVERSION		VER_STRING2
#define APPCOPYRIGHT	VER_LEGAL_COPYRIGHT

#define MESSAGEBOXW(x)   MessageBoxW(NULL, x, APPNAME, MB_OK|MB_ICONERROR);
#define MESSAGEBOXA(x)   MessageBoxA(NULL, x, APPNAME, MB_OK|MB_ICONERROR);

#define DEFAULT_DESKTOP "winsta0\\default"

int APIENTRY WinMain(HINSTANCE hInstance, HINSTANCE hPrevInstance, LPSTR lpCmdLine, int nCmdShow)
{
	D(("PESetEnv start"));

#ifdef _DEBUG_ALLOC
//    InitAllocCheck();
#endif

	int len = wcslen (GetCommandLineW()) + 1;
	LPSTR arg = (LPSTR) malloc(len * sizeof(char));
	int argc;
	LPWSTR *argvw = CommandLineToArgvW(GetCommandLineW(), &argc);
	LPSTR *argv = (LPSTR *) malloc(argc * sizeof(char *));
	int i;
	for (i = 0; i < argc; ++i)
	{
		int len2 = WideCharToMultiByte(CP_ACP, 0, argvw[i], -1, arg, len, NULL, NULL);
		argv[i] = (LPSTR) malloc((len2 + 1) * sizeof(char));
		strcpy(argv[i], arg);
	}
	free (arg);

	opterr = 0;
	PTCHAR pszStartingDirectory = NULL;
	int c;
	int len3;
	while ((c = getopt(argc, argv, _T("vd:"))) != EOF)
    {
        switch (c)
        {
            case _T('d'):
				if (pszStartingDirectory)
					free(pszStartingDirectory);
				len3 = ExpandEnvironmentStrings(optarg, NULL, 0) + 1;
				pszStartingDirectory = (PTCHAR) malloc(len3 * sizeof(char));
				if (!ExpandEnvironmentStrings(optarg, pszStartingDirectory, len3))
				{
					MESSAGEBOXW(SystemMessageW(GetLastError()));
			        return GetLastError();
				}
                break;
            case _T('v'):
				MESSAGEBOXW(L"Copyright (c) 2005-2015, Ross Smith II. MIT Licensed.");
                break;
			default:
				// silently ignore invalid options
				//MESSAGEBOXW(L"Invalid Option");
				break;
        }
    }

	LPSTR szCmdLineTemp = (LPSTR) malloc(len * sizeof(char *));
	strcpy (szCmdLineTemp, "");

	for (; optind < argc; ++optind)
	{
		strcat(szCmdLineTemp, argv[optind]);
		if (optind < argc - 1)
			strcat(szCmdLineTemp, " ");
	}

	for (i = 0; i < argc; ++i)
	{
		free(argv[i]);
	}

	if (strlen(szCmdLineTemp) == 0)
	{
		free(szCmdLineTemp);
		if (pszStartingDirectory)
			free(pszStartingDirectory);
		return -2;
	}

    int len4 = ExpandEnvironmentStrings(szCmdLineTemp, NULL, 0) + 1;
	LPSTR szCmdLine = (LPSTR) malloc(len4 * sizeof(char *));

    if (!ExpandEnvironmentStrings(szCmdLineTemp, szCmdLine, len4))
	{
		MESSAGEBOXW(SystemMessageW(GetLastError()));
		free(szCmdLine);
		free(szCmdLineTemp);
		if (pszStartingDirectory)
			free(pszStartingDirectory);
        return GetLastError();
	}

	free(szCmdLineTemp);

	int r = UpdatePathVariable();
	if (r != ERROR_SUCCESS)
	{
		// ignore the error
		// free(szCmdLine);
		// return r;
	}

    STARTUPINFO si;
    PROCESS_INFORMATION pi;

    ZeroMemory(&pi, sizeof(PROCESS_INFORMATION));
    ZeroMemory(&si, sizeof(STARTUPINFO));
    si.cb = sizeof(STARTUPINFO);
    si.lpDesktop = DEFAULT_DESKTOP;

    DWORD rv = CreateProcess(NULL,             // No module name (use command line).
                      szCmdLine,        // Command line.
                      NULL,             // Process handle not inheritable.
                      NULL,             // Thread handle not inheritable.
                      FALSE,            // Set handle inheritance to FALSE.
                      0,                // No creation flags.
                      NULL,             // Use parent's environment block.
                      pszStartingDirectory,	// starting directory.
                      &si,              // Pointer to STARTUPINFO structure.
                      &pi);             // Pointer to PROCESS_INFORMATION structure.

	free (szCmdLine);
	if (pszStartingDirectory)
		free(pszStartingDirectory);

	if (!rv)
	{
		MESSAGEBOXW(SystemMessageW(GetLastError()));
        return GetLastError();
	}

    DWORD exitcode = (DWORD) -1;

	WaitForSingleObject(pi.hProcess, INFINITE);
	GetExitCodeProcess(pi.hProcess, &exitcode);
    CloseHandle(pi.hProcess);
	CloseHandle(pi.hThread);

	D(("PESetEnv returning %d", exitcode));

    return exitcode;
}


UpdatePathVariable.cpp
// Copyright (c) 2005-2015 Ross Smith II. See Mit LICENSE in /LICENSE

#ifdef _MSC_VER
#pragma warning(disable:4996)
#endif


#define _WIN32_WINNT 0x501

#ifndef WIN32_LEAN_AND_MEAN
# define WIN32_LEAN_AND_MEAN 1
#endif

#include <windows.h>
#include <stdio.h>
#include <stdlib.h>

#include "UpdatePathVariable.h"
#include "debug.h"
#include "SystemMessage.h"

#define MAX_VALUE_LEN  32767

#define POSTFIXPATHREGKEY	L"Software\\PESetEnv\\Path\\Postfix"
#define ENVREGKEY       L"System\\CurrentControlSet\\Control\\Session Manager\\Environment"

#define APPNAME         L"PESetEnv"

#define MESSAGEBOXW(x)   MessageBoxW(NULL, x, APPNAME, MB_OK|MB_ICONERROR);
#define MESSAGEBOXA(x)   MessageBoxA(NULL, x, APPNAME, MB_OK|MB_ICONERROR);

#define BROADCAST_TIMEOUT (5000)

int UpdatePathVariable(void)
{
    HKEY hKey = NULL;
    PWCHAR szPath = NULL;
    PWCHAR szNewPath = NULL;
    PWCHAR szExpandedPath = NULL;
	DWORD rv = (DWORD) -1;

	D(("UpdatePathVariable() start"));

	for (;;)
	{
		WCHAR szVarName[MAX_VALUE_LEN];
		WCHAR szVarValue[MAX_VALUE_LEN];
		DWORD dwPathLen = 0;
		DWORD dwLen = 0;
		DWORD dwType;
		DWORD dwNewLen = 0;
		DWORD i = 0;

		rv = RegOpenKeyExW(HKEY_LOCAL_MACHINE, POSTFIXPATHREGKEY, 0, KEY_READ | KEY_QUERY_VALUE, &hKey);
		if (rv == ERROR_FILE_NOT_FOUND) {
			return ERROR_SUCCESS;
		}
		if (rv != ERROR_SUCCESS) {
			MESSAGEBOXW(SystemMessageW(rv));
			break;
		}

		for (;;)
		{
			dwLen = MAX_VALUE_LEN;
			rv = RegEnumValueW(hKey, i, szVarName, &dwLen, NULL, NULL, NULL, NULL);
			if (rv == ERROR_NO_MORE_ITEMS)
				break;
			if (rv != ERROR_SUCCESS)
			{
				MESSAGEBOXW(SystemMessageW(rv));
				break;
			}

			D(("dwPathLen=%5d, i=%5d, dwLen=%5d, szVarName=%S", dwPathLen, i, dwLen, szVarName));

			dwLen = MAX_VALUE_LEN;
			rv = RegQueryValueExW(hKey, szVarName, NULL, &dwType, (LPBYTE) szVarValue, &dwLen);
			if (rv == ERROR_NO_MORE_ITEMS)
				break;
			if (rv != ERROR_SUCCESS)
			{
				MESSAGEBOXW(SystemMessageW(rv));
				break;
			}

			D(("szVarValue=%S dwLen=%s", szVarValue, dwLen));

       		++i;

			if (dwLen == 0)
				continue;

			if (dwPathLen)
				++dwPathLen;
       		dwPathLen += dwLen;
		}
		if (rv != ERROR_NO_MORE_ITEMS && rv != ERROR_SUCCESS)
			break;

		D(("dwPathLen=%5d, i=%5d", dwPathLen, i));

		szPath = (PWCHAR) malloc(dwPathLen);
		if (szPath == NULL)
		{
			MESSAGEBOXW(SystemMessageW(rv));
		    break;
		}

		wcscpy(szPath, L"");

		i = 0;
		for (;;)
		{
	        dwLen = MAX_VALUE_LEN;
			rv = RegEnumValueW(hKey, i, szVarName, &dwLen, NULL, NULL, NULL, NULL);
			if (rv == ERROR_NO_MORE_ITEMS)
				break;
			if (rv != ERROR_SUCCESS)
			{
				MESSAGEBOXW(SystemMessageW(rv));
			    break;
			}

			D(("dwPathLen=%5d, i=%5d, szVarName=%S", dwPathLen, i, szVarName));

			dwLen = MAX_VALUE_LEN;
			rv = RegQueryValueExW(hKey, szVarName, NULL, &dwType, (LPBYTE) szVarValue, &dwLen);
			if (rv == ERROR_NO_MORE_ITEMS)
				break;
			if (rv != ERROR_SUCCESS)
			{
				MESSAGEBOXW(SystemMessageW(rv));
			    break;
			}

			D(("dwPathLen=%5d, szVarValue=%S szPath=%S", dwPathLen, szVarValue, szPath));

			++i;

			if (dwLen == 0)
				continue;

			if (wcslen(szPath))
				wcscat(szPath, L";");

			wcscat(szPath, szVarValue);
		}
		if (rv != ERROR_NO_MORE_ITEMS && rv != ERROR_SUCCESS)
			break;

		RegCloseKey(hKey);
		hKey = NULL;

		if (dwPathLen == 0)
		{
			// no paths need to be added
			rv = ERROR_SUCCESS;
			break;
		}

		rv = RegOpenKeyExW(HKEY_LOCAL_MACHINE, ENVREGKEY, 0, KEY_READ | KEY_WRITE | KEY_QUERY_VALUE, &hKey);
		if (rv != ERROR_SUCCESS)
		{
			MESSAGEBOXW(SystemMessageW(rv));
			break;
		}

		dwLen = MAX_VALUE_LEN;
		dwType = REG_SZ;
		rv = RegQueryValueExW(hKey, L"Path", NULL, &dwType, (LPBYTE) szVarValue, &dwLen);
		if (rv != ERROR_SUCCESS) {
			MESSAGEBOXW(SystemMessageW(rv));
			break;
		}

		dwNewLen = dwPathLen + dwLen + 4;
		D(("dwNewLen=%d", dwNewLen));

		szNewPath = (PWCHAR) malloc(dwNewLen + 4);
		if (szNewPath == NULL)
		{
			MESSAGEBOXW(SystemMessageW(GetLastError()));
		    break;
		}

		wcscpy(szNewPath, L"");

		if (dwLen)
		{
			wcscpy(szNewPath, szVarValue);
			wcscat(szNewPath, L";");
		}

		D(("szNewPath=%S", szNewPath));

		wcscat(szNewPath, szPath);

		D(("szNewPath=%S", szNewPath));

		szExpandedPath = (PWCHAR) malloc(32767);
		if (szExpandedPath == NULL)
		{
			MESSAGEBOXW(SystemMessageW(GetLastError()));
		    break;
		}
		if (!ExpandEnvironmentStringsW(szNewPath, szExpandedPath, 32767)) {
			MESSAGEBOXW(SystemMessageW(GetLastError()));
		    return GetLastError();
		}

		// todo only update path if running in PE
		if (dwType == REG_EXPAND_SZ)
		{
			rv = RegSetValueExW(hKey, L"Path", 0, dwType, (LPBYTE) szNewPath, wcslen(szNewPath) * sizeof(WCHAR));
		} else {
			rv = RegSetValueExW(hKey, L"Path", 0, dwType, (LPBYTE) szExpandedPath, wcslen(szExpandedPath) * sizeof(WCHAR));
		}

		if (rv != ERROR_SUCCESS)
		{
			MESSAGEBOXW(SystemMessageW(rv));
			break;
		}

		if (!SetEnvironmentVariableW(L"Path", szExpandedPath))
		{
			MESSAGEBOXW(SystemMessageW(GetLastError()));
			break;
		}

		static DWORD dummy;
		if (!SendMessageTimeout(HWND_BROADCAST, WM_SETTINGCHANGE, 0, (LPARAM) "Environment", SMTO_ABORTIFHUNG, BROADCAST_TIMEOUT, &dummy))
		{
			MESSAGEBOXW(SystemMessageW(GetLastError()));
			break;
		}

		rv = ERROR_SUCCESS;
		break;
	}

	if (szPath)
   		free(szPath);

	if (szNewPath)
   		free(szNewPath);

	if (szExpandedPath)
   		free(szExpandedPath);

	if (hKey)
		RegCloseKey(hKey);

	D(("UpdatePathVariable() returning %d", rv));

    return rv;
}